File: projects/asteroids/scripts/Bullet.lua
BlankE.addEntity("Bullet")

Bullet.net_sync_vars = {'speed','direction'}

function Bullet:init()
	self.speed = 800
	self:addShape("main","circle",{0,0,1})
	Timer(1.5):after(function() self:destroy() end):start()
	
	Net.addObject(self)
end

function Bullet:update(dt)
	if self.x > game_width then self.x = 0 end
	if self.x < 0 then self.x = game_width end
	if self.y > game_height then self.y = 0 end
	if self.y < 0 then self.y = game_height end
	
	self.onCollision["main"] = function(other, sep_vec)
		if other.parent.classname == "Asteroid" then
			if not self.net_object then
				Signal.emit('score',other.parent.points)
			end
			
			other.parent:hit(other.parent.direction)
			self:destroy()
		end
	end
end

function Bullet:draw()
	Draw.setColor("white")
	Draw.circle("fill",self.x,self.y,1)
end

File: projects/asteroids/scripts/Bullet.js
class Bullet extends Entity {
    init (player) {
		this.img = new Draw()
		this.x = player.x
		this.y = player.y
		this.moveDirection(player.sprite_angle-90, 10)
		Timer.after(1500, ()=> { this.destroy() })
    }
    update (dt) {
		this.img.draw(
			['lineStyle',1,Draw.white],
			['fill',Draw.white],
			['rect',-1,-1,2,2]
		)
		this.img.x = this.x;
		this.img.y = this.y;
		
		// wrapping
		if (this.y < 0) this.y = Game.height
		if (this.x < 0) this.x = Game.width
		if (this.y > Game.height) this.y = 0
		if (this.x > Game.width) this.x = 0
    }
	onDestroy () {
		this.img.destroy() 
	}
}
